### PR TITLE
Express ArrayList in ArrayListUnmanaged operations

### PR DESCRIPTION
**Pull request by Snektron**
I came across several issues in ArrayListUnmanaged, namely:
- An error in initCapacity which was introduced in #6232 https://github.com/ziglang/zig/blob/7d487a41627658bfda37f1f06f8e453c2b576b9c/lib/std/array_list.zig#L374
- An error in replaceRange introduced in #5511 https://github.com/ziglang/zig/blob/7d487a41627658bfda37f1f06f8e453c2b576b9c/lib/std/array_list.zig#L422

This PR changes the implementation of ArrayList to use that of ArrayListUnmanaged wherever possible, which ensures that the implementation of ArrayListUnmanaged and ArrayList are the same. The original test cases now also test ArrayListUnmanaged. Furthermore, this reduces code duplication in the standard library.

Note that i've changed `ArrayList.deinit` to require `*Self` like other std containers. I've changed constness of ArrayList wherever required.

Currently, most calls to ArrayListUnmanaged are called like
```zig
var unmanaged = self.toUnmanaged();
defer self.* = unmanaged.toManaged(self.allocator);
return unmanaged.addOne(self.allocator);
```
I hope that this is optimization/result location friendly, i haven't done any benchmarking.

### CODE REVIEW

File: lib/std/array_list.zig [hunk 1]
```diff
@@ -45,18 +46,15 @@ pub fn ArrayListAligned(comptime T: type, comptime alignment: ?u29) type {
...
-
-            return self;
+            var unmanaged = try Unmanaged.initCapacity(allocator, num);
+            return unmanaged.toManaged(allocator);
         }
 
         /// Release all allocated memory.
-        pub fn deinit(self: Self) void {
-            self.allocator.free(self.allocatedSlice());
+        pub fn deinit(self: *Self) void {
```
Comment (marler8997) [inline lib/std/array_list.zig:54]: Why the change to `*Self`?

Comment (Snektron) [reply lib/std/array_list.zig:54]: To be able to set `self` to `undefined` after deinitializing, which can cause bugs in debug mode. Note that this reflects the other std containers, which also take `*Self` in `deinit()`. `ArrayListAlignedUnmanaged` already did, but `ArrayListAligned` not yet.

Comment (Snektron) [reply lib/std/array_list.zig:54]: There are many places where constness needs to be changed though, i'm unsure if i should continue with that.

Comment (marler8997) [reply lib/std/array_list.zig:54]: > To be able to set self to undefined after deinitializing,

I see.  I wonder if we should provide a way to do one without the other?  Applications may want to de-allocate the array memory, but may not need or want to "de-initialize" the fields of the array structure.

Here's one example.  If you have a struct that contains an `ArrayListAligned` struct as a field, and it also wants to invalid its memory, then it's going to invalidate the same memory twice:
```zig
struct {
    a: ArrayListAligned,
    // more fields
    pub fn deinit(self: *@This()) {
        self.a.deinit();
        // .. deinitialize more fields
        self.* = undefined;  // we deinitialize the same self.a memory again
    }
}
```

In order to handle cases like these, our data structures need to provide a way to "deinit" their resources without invalidating their contents.  I think it's fine to have `deinit` do both by default, but what if we also expose deallocation of resources without invalidation as well, so maybe `deinitNoInvalidate`?

Then we can do:

```zig
struct {
    a: ArrayListAligned,
    // more fields
    pub fn deinit(self: *@This()) {
        self.a.deinitNoInvalidate(); // we invalidate everything at the end
        // .. deinitialize more fields
        self.* = undefined;
    }
}
```

Comment (marler8997) [reply lib/std/array_list.zig:54]: P.S. sorry my comments are mostly orthogonal to this PR, maybe I should open an issue for this?

Comment (Snektron) [reply lib/std/array_list.zig:54]: > Applications may want to de-allocate the array memory, but may not need or want to "invalidate" the fields of the array structure.
I think that `deinit` should always be destroying, and containers should provide separate methods for deallocating resources, if supported. This is already the case in many standard library containers, for example ArrayList(Unmanaged) have `.shrink` and `HashMap` has `clearAndFree()`. 

> P.S. sorry my comments are mostly orthogonal to this PR, maybe I should open an issue for this?
I think that would be best.

Comment (marler8997) [reply lib/std/array_list.zig:54]: Ok here it is: https://github.com/ziglang/zig/issues/6299

File: lib/std/array_list.zig [hunk 2]
```diff
@@ -24,14 +24,15 @@ pub fn ArrayListAligned(comptime T: type, comptime alignment: ?u29) type {
     }
     return struct {
         const Self = @This();
+        pub const Unmanaged = ArrayListAlignedUnmanaged(T, alignment);
 
         /// Content of the ArrayList
         items: Slice,
         capacity: usize,
         allocator: *Allocator,
```
Comment (Vexu) [inline lib/std/array_list.zig:32]: Should this be changed to
```zig
        unmanaged: Unmanaged,
        allocator: *Allocator,
```
to make it more consistent with `ArrayHashMap` and `HashMap`?

**Comment by daurnimator** [comment]
> Currently, most calls to ArrayListUnmanaged are called like
> 
> ```zig
> var unmanaged = self.toUnmanaged();
> defer self.* = unmanaged.toManaged(self.allocator);
> return unmanaged.addOne(self.allocator);
> ```
> 
> I hope that this is optimization/result location friendly, i haven't done any benchmarking.

My intuition is that this wouldn't be well optimized; e.g. you'd end up writing `self.allocator` needlessly each operation.

**Comment by Snektron** [comment]
After discussing with Andrew a few days ago, it was decided that ArrayList/ArrayListUnmanaged should be one of the few places where code is copy pasted. The argumentation was that this yields a smaller call stack, which is nice considering these types are used fairly fundamentally, and that the implementation of these types is fairly self-contained anyway. 

So instead of my solution, there should be some generic testing function (although i'm not quite sure how that would be implemented best). One idea is to have a mixin for the operations that don't have any allocations, although i don't think there are very many.

Thoughts?

File: lib/std/array_list.zig (hunk 2 above)
Comment (Snektron) [reply lib/std/array_list.zig:32]: I agree that this would be the most elegant solution, if the fields of the array list weren't part of the interface. Users are encouraged to directly refer to items in the ArrayList by using `.items[index]`, and changing this would be a big breaking change.

Comment (marler8997) [reply lib/std/array_list.zig:32]: This along with other use cases makes me think it might be useful for Zig to have a way to declare aliases.  For example:
```zig
    pub alias items = unmanaged.items;
```
This would mean code references `array_list.items` it would be equivalent to referencing `array_list.unmanaged.items`.

**Comment by Snektron** [comment]
Closing in favor of #6457
